### PR TITLE
[HAPROXY] JWT verification

### PR DESCRIPTION
**Pull request by ELDiablO59152**
### Added
- [API_PARSER] [SIGNALSCIENCES_NGWAF] New collector
- [HAPROXY] [CONF] JWT verification through workflow
### Fixed
- [CONFIG] [SYSTEM] Correct issue preventing some certificates to be deleted from system
- [OPENID] [MODELS] Get correct URLs for OpenID IDPs

### CODE REVIEW

File: vulture_os/portal/views/logon.py [hunk 1]
```diff
@@ -566,31 +568,43 @@ def openid_userinfo(request, portal_id=None, workflow_id=None):
...
+                if jwt_unverified:
+                    logger.debug(f"PORTAL::openid_userinfo: Token is a valid JWT {jwt_unverified}")
+                    jwt_iss = jwt_unverified["iss"]
+                    for connector in OpenIDRepository.objects.filter(enable_jwt=True, provider_url=jwt_iss):
+                        try:
+                            jwt_verified = jwt.decode(jwt=token,
+                                                    algorithms=[connector.jwt_signature_type],
+                                                    key=connector.jwt_key,
+                                                    issuer=connector.provider_url,
+                                                    audience=f"{scheme}://{fqdn}/" if connector.jwt_validate_audience else jwt_unverified['aud']
```
Comment (frikilax) [inline vulture_os/portal/views/logon.py:588]: no audience validation here, even if activated: this is just a compatiblity feature

File: vulture_os/portal/views/logon.py [hunk 2]
```diff
@@ -566,31 +568,43 @@ def openid_userinfo(request, portal_id=None, workflow_id=None):
...
+                                                    )
+                        except Exception as e:
```
Comment (frikilax) [inline vulture_os/portal/views/logon.py:590]: no generic exceptions when you can avoid them!
use https://pyjwt.readthedocs.io/en/stable/api.html?highlight=decode#exceptions

File: vulture_os/portal/views/logon.py [hunk 3]
```diff
@@ -566,31 +568,43 @@ def openid_userinfo(request, portal_id=None, workflow_id=None):
...
+                            logger.exception(f"PORTAL::openid_userinfo: Fail to verify JWT : {e}")
+                        else:
+                            logger.debug(f"PORTAL::openid_userinfo: JWT verified")
+                            jwt_verified['scope'].update({'exp': jwt_verified['exp'], 'iat': jwt_verified['iat']})
+                            ret = jwt_verified['scope']
```
Comment (frikilax) [inline vulture_os/portal/views/logon.py:595]:
```suggestion
                            ret = jwt_verified['scope']
                            ret.update({'exp': jwt_verified['exp'], 'iat': jwt_verified['iat']})
```

File: vulture_os/portal/views/logon.py [hunk 4]
```diff
@@ -566,31 +568,43 @@ def openid_userinfo(request, portal_id=None, workflow_id=None):
...
-            if jwt_unverified:
-                logger.debug("PORTAL::openid_userinfo: Token is a valid JWT")
-                jwt_iss = jwt_unverified["iss"]
-                for connector in OpenIDRepository.objects.filter(enable_jwt=True, provider_url=jwt_iss):
-                    jwt_verified = jwt.decode(jwt=token, algorithms=[connector.jwt_signature_type], key=connector.jwt_key, issuer=connector.provider_url)
-                    if jwt_verified:
-                        logger.debug(f"PORTAL::openid_userinfo: JWT verified")
-                        ret = jwt_verified
-                        break
+        if len(token.split('.')) == 3:
```
Comment (frikilax) [inline vulture_os/portal/views/logon.py:571]: no, keep try except method to fallback to regular token validation

File: vulture_os/portal/views/logon.py [hunk 5]
```diff
@@ -566,31 +568,43 @@ def openid_userinfo(request, portal_id=None, workflow_id=None):
...
             assert session.exists(), "Session not found."
             assert session['scope'], "Session does not contain any scope."
             # Add internal Oauth2 attributes
-            session['scope'].update({'exp': session['exp']})
-            session['scope'].update({'iat': session['iat']})
-            ret = session["scope"]
+            session['scope'].update({'exp': session['exp'], 'iat': session['iat']})
+            ret = session['scope']
 
+        assert ret != {}, "Authentication failure"
```
Comment (frikilax) [inline vulture_os/portal/views/logon.py:607]: useless when previous double-validation was used: first JWt, then regular if no JWT was found
regular handles assertions and failure cases

File: vulture_os/workflow/config/haproxy_portal.conf [hunk 6]
```diff
@@ -122,35 +120,47 @@ frontend Workflow_{{conf.id}}
...
+    http-request set-var(txn.aud) http_auth_bearer,jwt_payload_query('$.aud')
+    http-request set-var(txn.exp) http_auth_bearer,jwt_payload_query('$.exp','int')
+    http-request set-var(txn.now) date()
+
+    {% for repo in conf.authentication.openid_repos if repo.enable_jwt -%}
+    # Validate JWT for repo {{repo.name}}
+    http-request set-var(txn.valid_jwt) bool(true) if { var(txn.alg) -m str "{{repo.jwt_signature_type}}" } { var(txn.iss) -m str "{{repo.provider_url}}" } {% if repo.jwt_validate_audience -%} { var(txn.aud) -m str "{{conf.redirect_uri}}" } { http_auth_bearer,jwt_verify(txn.alg,"{{repo.jwt_key if repo.jwt_signature_type in ("HS256", "HS384", "HS512") else repo.get_jwt_key_filename()}}") -m int 1 } {%- endif %} { var(txn.exp),sub(txn.now) -m int gt 0 }
+    {% endfor -%} {# for repo in conf.authentication.openid_repos if repo.enable_jwt #}
+
     # Check session with LUA
```
Comment (frikilax) [inline vulture_os/workflow/config/haproxy_portal.conf:138]: needs to be rewritten to validate all conditions at once: here conditions between different OpenID JWT parameters can be mixed

File: vulture_os/workflow/config/haproxy_portal.conf [hunk 7]
```diff
@@ -122,35 +120,47 @@ frontend Workflow_{{conf.id}}
...
     # Workflow {{conf.name}}
+
+    # get header part of the JWT
+    http-request set-var(txn.alg) http_auth_bearer,jwt_header_query('$.alg')
+
+    # get payload part of the JWT
+    http-request set-var(txn.iss) http_auth_bearer,jwt_payload_query('$.iss')
+    http-request set-var(txn.aud) http_auth_bearer,jwt_payload_query('$.aud')
+    http-request set-var(txn.exp) http_auth_bearer,jwt_payload_query('$.exp','int')
+    http-request set-var(txn.now) date()
```
Comment (frikilax) [inline vulture_os/workflow/config/haproxy_portal.conf:131]: need to add a jinja condition to calculate those values only when necessary (JWT validation activated on this workflow)

File: vulture_os/authentication/user_portal/models.py [hunk 8]
```diff
@@ -651,23 +651,21 @@ def to_html_template(self):
 
     def get_openid_callback_url(self, req_scheme, workflow_host, workflow_path, repo_id):
         if self.enable_external:
-            base_url = build_url("https" if self.external_listener.tls_profiles.count()>0 else "http", self.external_fqdn, self.external_listener.port)
+            base_url = build_url("https" if self.external_listener.has_tls() else "http", self.external_fqdn, '' if len(self.external_fqdn.split(':')) > 1 else self.external_listener.listener_set.first().port, '/')
```
Comment (frikilax) [inline vulture_os/authentication/user_portal/models.py:654]: not the right correction: port shouldn't be included in the external FQDN!

File: vulture_os/authentication/templates/authentication/openid_edit.html [hunk 9]
```diff
@@ -351,10 +342,13 @@ <h4 class="modal-title">{% translate "OpenID configuration" %}</h4>
...
-    toggle_jwt();
-  })
-  toggle_jwt();
+  /* Show jwt class if enable_jwt is True */
+  $('#id_enable_jwt').on('click', () => {
+    if( self.checked )
+      $('.jwt').show();
+    else
+      $('.jwt').hide();
+  }).trigger();
```
Comment (frikilax) [inline vulture_os/authentication/templates/authentication/openid_edit.html:352]: No! This doesn't work!

File: vulture_os/authentication/openid/views.py [hunk 10]
```diff
@@ -85,13 +88,28 @@ def edit(request, object_id=None, api=False):
...
         # If provider_url changed, force reload of configuration
         if "provider_url" in form.changed_data:
             repo.last_config_time = None
+        if "jwt_key" in form.changed_data and repo.jwt_signature_type not in ("HS256", "HS384", "HS512"):
+            Cluster.api_request("system.config.models.delete_conf", old_jwt_key_filename)
+            repo.save_conf()
         repo.save()
 
+        portal = UserAuthentication.objects.get(name=repo.name.replace("Connector_", ""))
+        portal.save_conf()
```
Comment (frikilax) [inline vulture_os/authentication/openid/views.py:103]: Careful! an OpenID Connector is not always linked to a Vulture IDP!

File: vulture_os/authentication/openid/views.py [hunk 11]
```diff
@@ -85,13 +88,28 @@ def edit(request, object_id=None, api=False):
...
     if request.method in ("POST", "PUT") and form.is_valid():
+        old_jwt_key_filename = repo.get_jwt_key_filename()
         # Save the form to get an id if there is not already one
         repo = form.save(commit=False)
         # If provider_url changed, force reload of configuration
         if "provider_url" in form.changed_data:
             repo.last_config_time = None
+        if "jwt_key" in form.changed_data and repo.jwt_signature_type not in ("HS256", "HS384", "HS512"):
+            Cluster.api_request("system.config.models.delete_conf", old_jwt_key_filename)
+            repo.save_conf()
```
Comment (frikilax) [inline vulture_os/authentication/openid/views.py:99]: Looks like 2 different situations and conditions to me:
1. need to delete the file if 
  - jwt_signature_type changed from HSxxx to something else
  - (OpenID is deleted)
  - OpenID name is modified (as name is included in filename)
2. need to update existing configuration file if
  - name is changed (see above)
  - jwt_key is changed

**Review by frikilax — CHANGES_REQUESTED** [review]

File: vulture_os/workflow/config/haproxy_portal.conf (hunk 7 above)
Comment (ELDiablO59152) [reply vulture_os/workflow/config/haproxy_portal.conf:131]: I guess this comment was written before the last commit as we discussed earlier

File: vulture_os/authentication/user_portal/models.py (hunk 8 above)
Comment (ELDiablO59152) [reply vulture_os/authentication/user_portal/models.py:654]: I understand what you mean but self.external_listener.port does not exist so there will be a problem if FQDN does not contain the port on IDP portals

File: vulture_os/authentication/openid/views.py (hunk 10 above)
Comment (ELDiablO59152) [reply vulture_os/authentication/openid/views.py:103]: You're right ! I'll add an if exists statement

File: vulture_os/authentication/openid/models.py [hunk 12]
```diff
@@ -249,11 +250,42 @@ def to_html_template(self):
             'callback_url': f"/oauth2/callback/{self.id_alea}"
         }
 
+    def get_jwt_key_filename(self):
+        return f"/var/db/pki/{self.name}-{self.id}.pub"
```
[the remark this answers is not in the document]
Comment (ELDiablO59152) [reply vulture_os/authentication/openid/models.py]: Changed

File: vulture_os/portal/views/logon.py (hunk 1 above)
Comment (ELDiablO59152) [reply vulture_os/portal/views/logon.py]: Changed with options: verify_aud=false

File: vulture_os/portal/views/logon.py (hunk 2 above)
Comment (ELDiablO59152) [reply vulture_os/portal/views/logon.py]: Exception catching improved

File: vulture_os/portal/views/logon.py (hunk 4 above)
Comment (ELDiablO59152) [reply vulture_os/portal/views/logon.py]: Done

File: vulture_os/portal/views/logon.py (hunk 5 above)
Comment (ELDiablO59152) [reply vulture_os/portal/views/logon.py]: Line deleted

File: vulture_os/workflow/config/haproxy_portal.conf (hunk 7 above)
Comment (ELDiablO59152) [reply vulture_os/workflow/config/haproxy_portal.conf:134]: The solution I provided doesn't look very nice to me, feel free to suggest something else

File: vulture_os/authentication/templates/authentication/openid_edit.html (hunk 9 above)
Comment (ELDiablO59152) [reply vulture_os/authentication/templates/authentication/openid_edit.html]: I fixed this issue

File: vulture_os/authentication/openid/views.py (hunk 10 above)
Comment (ELDiablO59152) [reply vulture_os/authentication/openid/views.py]: Conf files refreshing of the related IDP is no longer needed

File: vulture_os/authentication/openid/views.py (hunk 11 above)
Comment (ELDiablO59152) [reply vulture_os/authentication/openid/views.py]: The conditions above are now taken into account

**Pull request merged**